### ISSUE TITLE
自实现Videx-stat-server

### ISSUE THREAD

**Issue by spenxu** [bytedance/videx#8]
HI， 我目前想基于springboot 实现Videx-stat-server， 查看接口/ask_videx 发现optimizer 层主要通过Http 获取如下几个统计信息：
![Image](https://github.com/user-attachments/assets/88b8df08-c4d8-4535-8c5d-e7ecca311fa8)

能帮忙解释下这几个函数具体各自含义吗？盼回复！谢谢

**Comment by spenxu** [comment]
另一个关于records in range 评估的准确性的疑问？ 
假设有条件： where c > 10 and c<100 and d > 10 , 这里涉及到两列c、d的选择性
但是我们的统计信息只有单列的ndv 和 直方图， 如果简单的认为：

_**c,d 的选择性 =  c 的选择性 * d的选择性**_ 

肯定是不准确的， 可能c和d 的值存在相关性

请问这里有什么其他的思考吗？

**Comment by kr11** [comment]
> HI， 我目前想基于springboot 实现Videx-stat-server， 查看接口/ask_videx 发现optimizer 层主要通过Http 获取如下几个统计信息： ![Image](https://github.com/user-attachments/assets/88b8df08-c4d8-4535-8c5d-e7ecca311fa8)
> 
> 对于这几个统计方法的描述不知道我理解的对不对？
> 
> 1. table_model.scan_time： 实际计算的是如果全表扫描，花费的代价是多少
> 2. table_model.get_memory_buffer_size：这个很直白， 获取memory_buffer_size大小，估计是用来评估JOIN 或者创建零时表的成本
> 3. table_model.records_in_range：这个应该是核心，根据条件和我们的ndv+直方图，返回预估满足条件的行数
> 4. table_model.info_low：返回表的一些基础信息
> 
> Videx 从server获取信息是不是只需要这几个信息？这对我springboot实现很重要，盼回复！谢谢：）

hi @spenxu 

Your observation is correct; VIDEX-optimizer sends all requests to the Statistic-Server via ask_videx, specifically including the following types:
1. `table_model.info_low`: Returns basic information about the table, including `table_rows`, `data_file_length`, `ndv`, etc. See the example in `test/videx/test_info_low.py`:

```markdown
expect_gt = [
    200,
    "OK",
    {
        'stat_n_rows': 5799239,
        "stat_clustered_index_size": 57152,
        'stat_sum_of_other_index_sizes': 69962,
        "data_file_length": 936378368,
        'index_file_length': 1146257408,
        "data_free_length": 2097152,
        'pct_cached #@# LINEITEM_FK1': 1,
        'rec_per_key #@# LINEITEM_FK1 #@# L_ID': 1.0,
    }
]
```

2. `table_model.scan_time`: A simple function that essentially equals `clustered_index_size`.
3. `table_model.get_memory_buffer_size`: A simple function that essentially equals `innodb_buffer_pool_size`.
4. `table_model.records_in_range`: The cardinality algorithm. Although request is in format of JSON, the format definition is a little hard-to-understand. Strongly consider referring to `VidexModelBase.records_in_range`.

Additionally, a few suggestions:
1. `videx_statistic_server` also has a `create_task_meta` interface for initialization, which accepts various data types. However, this interface is not initiated by videx-optimizer, and you can implement it freely.
2. The logic of each method can be referenced in `src/sub_platforms/sql_opt/videx/model/` under `videx_strategy.py` and `videx_model_innodb.py`.
3. The input and return formats for various requests can be seen in `test/videx/test_info_low.py` and `test/videx/test_records_in_range.py`.
4. A good practice is to start a Python server as a control group and use TPCH SQLs to trigger the complete process. The Python server will print out complete input and output data (in JSON format).

---
hi @spenxu 

你的观察是正确的，VIDEX-optimizer 会向 Statistic-Server发送的请求全都会经过 ask_videx，细分来说就是这几种：
1. table_model.info_low：返回表的一些基础信息，包括 table_rows，data_file_length、ndv 等等。样例可以参考 `test/videx/test_info_low.py`：
2. table_model.scan_time：一个较为简单的函数，基本等于 `clustered_index_size`
3. table_model.get_memory_buffer_size：一个较为简单的函数，基本等于 `innodb_buffer_pool_size`
4. table_model.records_in_range：也就是 cardinality 算法。虽然已经转化为 json request，但转换规则仍然是较为复杂的，这里强烈建议参考 `VidexModelBase.records_in_range`。

此外，还有一些建议：
1. videx_statistic_server 还会有一个 create_task_meta 接口来进行初始化，接收各种数据。但这个接口不是由 videx-optimizer 发起的，你可以自由实现。
2. 各种方法的实现逻辑可以参考 `src/sub_platforms/sql_opt/videx/model/` 下面的 `videx_strategy.py` 和 `videx_model_innodb.py`；
3. 各种请求的输入和返回格式可以参考 `test/videx/test_info_low.py` 和 `test/videx/test_records_in_range.py`
4. 一种良好的实践是，启动一个 python-server 作为对照组，并使用 TPCH sqls 来触发完整的流程。python-server 会打印出完整的输入和输出数据（json 格式）。

**Comment by kr11** [comment]
> 另一个关于records in range 评估的准确性的疑问？ 假设有条件： where c > 10 and c<100 and d = 10 , 这里涉及到两列c、d的选择性 但是我们的统计信息只有单列的ndv 和 直方图， 如果简单的认为：
> 
> _**c,d 的选择性 = c 的选择性 * d的选择性**_
> 
> 肯定是不准确的， 可能c和d 的值存在相关性
> 
> 请问这里有什么其他的思考吗？

Great question! VIDEX, or any what-if analysis tool, faces major challenges in estimating the joint distribution of multiple columns, including their NDVs (Number of Distinct Values) and cardinalities. This is a hot topic in the database research area (learned cardinality and learned NDV). Here are some suggestions:

1. VIDEX collects multi-column NDV information from existing indexes, such as index(c1, c2, c3), from which we can derive NDV(c1, c2) and NDV(c1, c2, c3).
2. We only collect single-column NDVs and histograms because traversing all combinations is impractical (O(N) -> O(2^N)). Based on single-column data, we estimate the joint distribution under the assumption of independence, which, although sometimes biased, is acceptable in many scenarios.
3. Our team plans to gradually open-source our dataless multi-column NDV & cardinality solutions, such as those discussed in our VLDB 25 paper (accepted): PLM4NDV: Minimizing Data Access for Number of Distinct Values Estimation with Pre-trained Language Models. You can refer to our paper here: [PLM4NDV Paper](https://arxiv.org/abs/2504.00608).
4. Most importantly, we look forward to VIDEX serving as an experimental platform to help researchers and developers find more practical algorithms for multi-column NDV and cardinality estimation.


---

非常好的问题！VIDEX ，或者说所有 what-if analysis 最大的难点就是估计多列联合分布，包括多列 ndv 和多列 cardinality。这也是database 研究领域非常热门的话题（learned cardinality and learned ndv）。这里有一些建议：

1. VIDEX 采集了现有索引的多列 ndv 信息，例如 index(c1,c2,c3)，我们可以拿到 ndv(c1, c2) 和 ndv(c1, c2, c3)
2. 我们仅采集单列 ndv 和直方图，是因为遍历所有组合是不可承受的（O(N) -> O(2^N))。基于单列信息，采用独立假设估算联合分布，虽然会有偏差，但是在很多场景下也可以接受。
3. 我们团队会逐步开源我们的 dataless multi-column ndv & cardinality，例如，你可以参考我们的 VLDB 25 paper (accepted): [PLM4NDV](https://arxiv.org/abs/2504.00608)
4. 最后也是最重要的，我们很期待 VIDEX 作为一个试验平台，帮助研究者和开发者找到更实用的多列 ndv 和多列 cardinality 算法。

**Comment by spenxu** [comment]
@kr11 
Thank you very much for your response！
 Additionally, I’m really looking forward to the release of a plugin version based on MySQL 5.7. I’ve noticed that, with the same structure and data, the execution plans in MySQL 8 and MySQL 5.7 show significantly different cost values—especially when self-joins or multi-table joins are involved, where the handling differs completely between 5.x and 8.x. I’m wondering if there are any plans regarding this?

**Comment by kr11** [comment]
> [@kr11](https://github.com/kr11) Thank you very much for your response！ Additionally, I’m really looking forward to the release of a plugin version based on MySQL 5.7. I’ve noticed that, with the same structure and data, the execution plans in MySQL 8 and MySQL 5.7 show significantly different cost values—especially when self-joins or multi-table joins are involved, where the handling differs completely between 5.x and 8.x. I’m wondering if there are any plans regarding this?

Yes, we have made significant progress on VIDEX supporting MySQL 5.7. Currently, we are testing VIDEX to simulate MySQL 5.7 using TPCH queries like Q21, Q2, Q3, etc., and have achieved **very high accuracy**.

Below is an example  comparing VIDEX with MySQL 5.7 on TPCH Q21:
![Image](https://github.com/user-attachments/assets/db577c8f-3df4-4943-bcf7-5bbb55f955a4)

The code will be updated within the next 1-2 weeks.

About MySQL 5.7, let's continue our discussion in issue #3

**Comment by kr11** [comment]
hi @spenxu we've added a new pull request #14 to support MySQL 5.7 in standalone mode and would greatly appreciate your feedback on TPC-H tiny or other workloads.

The readme has been updated, and the dockerhub image is kangrongme/videx:0.0.3.

**Comment by spenxu** [comment]
Wow, Great! Looking forward to more new features!

**Comment by spenxu** [comment]
> hi [@spenxu](https://github.com/spenxu) we've added a new pull request [#14](https://github.com/bytedance/videx/pull/14) to support MySQL 5.7 in standalone mode and would greatly appreciate your feedback on TPC-H tiny or other workloads.
> 
> The readme has been updated, and the dockerhub image is kangrongme/videx:0.0.3.



Wow, Great! Looking forward to more new features!
